src/logic/Window.swift

```diff
@@ -145,13 +183,43 @@ class Window {
         }
     }
 
+    func gotoSpace(_ spaceId: CGSSpaceID) {
+        CGSMoveWindowsToManagedSpace(cgsMainConnectionId, [App.app.helperWindow.windowNumber] as NSArray, spaceId)
+        App.app.showHelperWindow()
+    }
+
     func focus() {
         if isWindowlessApp {
             if let bundleID = application.runningApplication.bundleIdentifier {
                 NSWorkspace.shared.launchApplication(withBundleIdentifier: bundleID, additionalEventParamDescriptor: nil, launchIdentifier: nil)
             } else {
                 application.runningApplication.activate(options: .activateIgnoringOtherApps)
             }
+        } else if self.axUiElement == nil {
+            // the window is in other space, we can not get AXUIElement before we go to it's space
+            // when we want to focus the window, it means we want to go to the space and focus the window
+            // so we just go to that space by using helperWindow, then we get AXUIElement and focus the window
+            gotoSpace(self.spaceId)
+            do {
+                retryAxCallUntilTimeout { [weak self] in
+                    guard let self = self else { return }
+                    try self.getAxUiElementAndObserveEvents()
+                    if self.axUiElement == nil {
+                        // retry until we get axUiElement
+                        throw AxError.runtimeError
+                    }
+                    BackgroundWork.accessibilityCommandsQueue.asyncWithCap { [weak self] in
+                        guard let self = self else { return }
+                        var psn = ProcessSerialNumber()
+                        GetProcessForPID(self.application.pid, &psn)
+                        _SLPSSetFrontProcessWithOptions(&psn, self.cgWindowId, .userGenerated)
+                        self.makeKeyWindow(psn)
+                        self.axUiElement.focusWindow()
+                    }
+                }
+            } catch {
```

Comment:
I don't think exceptions can bubble up to here. `retryAxCallUntilTimeout` happens on another thread, and absorbs exceptions within